// @sanity/base/lib/__legacy/@sanity/components/popover/stories/moveRef.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MoveRefStory = MoveRefStory;

var _popover = require("part:@sanity/components/popover");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
function MoveRefStory() {
  var refIndex = (0, _knobs.select)('Reference element', {
    '0': 'First',
    '1': 'Second'
  }, '0', 'Debug');
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/dialogs/default",
    propTables: [_popover.Popover]
  }, /*#__PURE__*/_react.default.createElement(PopoverExample, {
    refIndex: Number(refIndex)
  })));
}

function PopoverExample(_ref) {
  var refIndex = _ref.refIndex;
  var ref1 = (0, _react.useRef)(null);
  var ref2 = (0, _react.useRef)(null);
  var open = true;
  var placement = 'bottom';
  var ref = refIndex === 0 ? ref1 : ref2;

  var popoverContent = /*#__PURE__*/_react.default.createElement("div", {
    style: {
      padding: '1em'
    }
  }, "popover content");

  return /*#__PURE__*/_react.default.createElement(_popover.Popover, {
    content: popoverContent,
    open: open,
    placement: placement,
    style: {
      transition: 'transform 200ms'
    },
    targetElement: ref.current
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("span", {
    ref: ref1,
    style: {
      display: 'inline-block',
      padding: '0.25em',
      background: 'rgba(127, 127, 127, 0.5)'
    }
  }, "reference 1"), /*#__PURE__*/_react.default.createElement("span", {
    ref: ref2,
    style: {
      display: 'inline-block',
      padding: '0.25em',
      background: 'rgba(127, 127, 127, 0.5)'
    }
  }, "reference 2")));
}
