// @sanity/portable-text-editor/lib/utils/selection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createKeyedPath = createKeyedPath;
exports.createArrayedPath = createArrayedPath;
exports.normalizeSelection = normalizeSelection;
exports.toPortableTextRange = toPortableTextRange;
exports.toSlateRange = toSlateRange;

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _slate = require("slate");

var _types = require("@sanity/types");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function createKeyedPath(point, editor) {
  var blockPath = [point.path[0]];
  var block;

  try {
    ;

    var _Editor$node = _slate.Editor.node(editor, blockPath, {
      depth: 1
    });

    var _Editor$node2 = _slicedToArray(_Editor$node, 1);

    block = _Editor$node2[0];
  } catch (err) {
    return null;
  }

  if (!block || !_slate.Element.isElement(block)) {
    return null;
  }

  var keyedBlockPath = [{
    _key: block._key
  }];

  if (editor.isVoid(block)) {
    return keyedBlockPath;
  }

  var keyedChildPath;
  var child;
  var childPath = point.path.slice(0, 2);

  if (childPath.length === 2) {
    try {
      ;

      var _Editor$node3 = _slate.Editor.node(editor, childPath, {
        depth: 2
      });

      var _Editor$node4 = _slicedToArray(_Editor$node3, 1);

      child = _Editor$node4[0];
    } catch (err) {
      return null;
    }

    keyedChildPath = ['children', {
      _key: child._key
    }];
  }

  return keyedChildPath ? [...keyedBlockPath, ...keyedChildPath] : keyedBlockPath;
}

function createArrayedPath(point, editor) {
  if (!editor) {
    return [];
  }

  var _ref = Array.from(_slate.Editor.nodes(editor, {
    at: [],
    match: n => (0, _types.isKeySegment)(point.path[0]) && n._key === point.path[0]._key
  }))[0] || [undefined, undefined],
      _ref2 = _slicedToArray(_ref, 2),
      block = _ref2[0],
      blockPath = _ref2[1];

  if (!block || !_slate.Element.isElement(block)) {
    return [];
  }

  if (editor.isVoid(block)) {
    return blockPath;
  }

  var childPath = [point.path[2]];
  var childIndex = block.children.findIndex(child => (0, _isEqual2.default)([{
    _key: child._key
  }], childPath));

  if (childIndex >= 0 && block.children[childIndex]) {
    var child = block.children[childIndex];

    if (_slate.Element.isElement(child) && editor.isVoid(child)) {
      return blockPath.concat(childIndex).concat(0);
    }

    return blockPath.concat(childIndex);
  }

  return blockPath;
}

function normalizePoint(point, value) {
  if (!point || !value) {
    return null;
  }

  var newPath = [];
  var newOffset = point.offset || 0;
  var blockKey = typeof point.path[0] === 'object' && '_key' in point.path[0] && point.path[0]._key;
  var childKey = typeof point.path[2] === 'object' && '_key' in point.path[2] && point.path[2]._key;
  var block = value.find(blk => blk._key === blockKey);

  if (block) {
    newPath.push({
      _key: block._key
    });
  } else {
    return null;
  }

  if (block && point.path[1] === 'children') {
    if (!block.children || block.children.length === 0) {
      return null;
    }

    var child = block.children.find(cld => cld._key === childKey);

    if (child) {
      newPath.push('children');
      newPath.push({
        _key: child._key
      });
      newOffset = child.text && child.text.length >= point.offset ? point.offset : child.text && child.text.length || 0;
    } else {
      return null;
    }
  }

  return {
    path: newPath,
    offset: newOffset
  };
}

function normalizeSelection(selection, value) {
  if (!selection || !value || value.length === 0) {
    return null;
  }

  var newAnchor = null;
  var newFocus = null;
  var anchor = selection.anchor,
      focus = selection.focus;

  if (anchor && value.find(blk => (0, _isEqual2.default)({
    _key: blk._key
  }, anchor.path[0]))) {
    newAnchor = normalizePoint(anchor, value);
  }

  if (focus && value.find(blk => (0, _isEqual2.default)({
    _key: blk._key
  }, focus.path[0]))) {
    newFocus = normalizePoint(focus, value);
  }

  if (newAnchor && newFocus) {
    return {
      anchor: newAnchor,
      focus: newFocus
    };
  }

  return null;
}

function toPortableTextRange(editor) {
  if (!editor.selection) {
    return editor.selection;
  }

  var anchor = null;
  var focus = null;
  var anchorPath = createKeyedPath(editor.selection.anchor, editor);

  if (anchorPath) {
    anchor = {
      path: anchorPath,
      offset: editor.selection.anchor.offset
    };
  }

  var focusPath = createKeyedPath(editor.selection.focus, editor);

  if (focusPath) {
    focus = {
      path: focusPath,
      offset: editor.selection.focus.offset
    };
  }

  var range = anchor && focus ? {
    anchor,
    focus
  } : null;
  return range;
}

function toSlateRange(selection, editor) {
  if (!selection || !editor) {
    return null;
  }

  var anchor = {
    path: createArrayedPath(selection.anchor, editor),
    offset: selection.anchor.offset
  };
  var focus = {
    path: createArrayedPath(selection.focus, editor),
    offset: selection.focus.offset
  };
  var range = anchor && focus ? {
    anchor,
    focus
  } : null;
  return range;
}
