// @sanity/base/lib/__legacy/@sanity/components/avatar/stories/stack.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StackStory = StackStory;

var _color = require("@sanity/color");

var _avatar = require("part:@sanity/components/avatar");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
var colors = [{
  dark: _color.hues.blue[400].hex,
  light: _color.hues.blue[500].hex
}, {
  dark: _color.hues.purple[400].hex,
  light: _color.hues.purple[500].hex
}, {
  dark: _color.hues.magenta[400].hex,
  light: _color.hues.magenta[500].hex
}, {
  dark: _color.hues.orange[400].hex,
  light: _color.hues.orange[500].hex
}, {
  dark: _color.hues.yellow[400].hex,
  light: _color.hues.yellow[500].hex
}, {
  dark: _color.hues.cyan[400].hex,
  light: _color.hues.cyan[500].hex
}];

function getRandomColor() {
  var idx = Math.floor(Math.random() * colors.length);
  return colors[idx];
}

function StackStory() {
  var props = {
    maxLength: (0, _knobs.number)('Max. length', 4, 'Props'),
    size: (0, _knobs.select)('Size', {
      small: 'Small',
      medium: 'Medium',
      large: 'Large'
    }, 'small', 'Props')
  };
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/dialogs/default",
    propTables: [_avatar.AvatarStack]
  }, /*#__PURE__*/_react.default.createElement(_avatar.AvatarStack, props, /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "cy"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ma"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "ye"
  }), /*#__PURE__*/_react.default.createElement(_avatar.Avatar, {
    color: getRandomColor(),
    initials: "bl"
  }))));
}
